GDJS/scripts/lib/build-esm-library-types.js

```diff
@@ -0,0 +1,140 @@
+const ts = require('typescript');
+const path = require('path');
+const { writeFile } = require('fs/promises');
+const { readFileSync } = require('fs');
+
+/** @param {string[]} str */
+const relativePath = (...str) => path.relative(...str).replace(/\\/g, '/');
+
+const gdevelopRootPath = path.join(__dirname, '..', '..', '..');
+const gdevelopTypesDirectory = path.join(
+  gdevelopRootPath,
+  'GDJS',
+  'Runtime',
+  'types',
+  'auto-generated'
+);
+
+/**
+ * Generates a bundled dts for a module using the TypeScript programmatic API.
+ * @param {string} pathToModule The path to the module to generate a dts from.
+ * @returns {Promise<[string, string]>} The DTS file contents as plain text.
+ */
+const generateDTS = (pathToModule) => {
+  const mainModuleDirectoryName = path.dirname(pathToModule);
+  const modulePrefix = path.basename(pathToModule, '.ts');
+  const expectedMainModuleOutputName = pathToModule
+    .replace('.ts', '.d.ts')
+    .replace(/\\/g, '/');
+
+  /**
+   * @param {string} dts
+   * @param {string} relativePath
+   */
+  const rewriteImports = (dts, dtsPath) => {
+    let rewrittenDTS = dts;
+
+    for (const match of dts.matchAll(
+      /((export|import) .*) from ["'](\.?\.\/.*)["']/gm
+    ))
+      rewrittenDTS = rewrittenDTS.replace(
+        match[0],
+        `${match[1]} from '${modulePrefix}/${relativePath(
+          mainModuleDirectoryName,
+          `${mainModuleDirectoryName}/${dtsPath}/${match[3]}`
+        )}'`
+      );
+
+    return rewrittenDTS;
+  };
+
+  let mainDTS = '';
+  let bundledSubModules = '';
+
+  const program = ts.createProgram([pathToModule], {
+    declaration: true,
+    emitDeclarationOnly: true,
+    typeRoots: [
+      gdevelopTypesDirectory,
+      path.join(gdevelopRootPath, 'GDJS', 'node_modules', '@types'),
+    ],
+  });
+  const { diagnostics } = program.emit(
+    undefined,
+    (fileName, dts) => {
+      if (fileName !== expectedMainModuleOutputName) {
+        const moduleName = path
+          .relative(mainModuleDirectoryName, fileName)
+          .replace(/\\/g, '/')
+          .replace('.d.ts', '');
+        const modulePath = path.dirname(moduleName);
+        const moduleFinalDTS = rewriteImports(dts, modulePath).replace(
+          /declare /gm,
+          ''
+        );
+        bundledSubModules += `declare module "${modulePrefix}/${moduleName}" {${moduleFinalDTS}};\n\n`;
```

Comment:
I don't know module terminology well, but `bundled` make me think of files merged together.
What do you think about `subModuleDeclarationsString`? (the "String" suffix avoid confusion with an iterable)

Reply:
The thing is, this actually is merging files together. It merges all imports of the index module into one declaration file that declares each of these modules.